GitCommands/Git/GitModule.cs

```diff
@@ -102,11 +102,77 @@ public sealed class GitModule : IGitModule
 
         public GitModule([CanBeNull] string workingDir)
         {
-            _superprojectInit = false;
             WorkingDir = (workingDir ?? "").EnsureTrailingPathSeparator();
             WorkingDirGitDir = GitDirectoryResolverInstance.Resolve(WorkingDir);
             _indexLockManager = new IndexLockManager(this);
             _commitDataManager = new CommitDataManager(() => this);
+
+            (SuperprojectModule, SubmodulePath, SubmoduleName) = TryInitSubmodule();
+
+            return;
+
+            (GitModule superprojectModule, string submodulePath, string submoduleName) TryInitSubmodule()
```

Comment:
Could you please provide clarity about this method's intentions. 
`TryXxx` methods typically have signatures `bool TryXxx(out T)` and have a companion method `Xxx` that may throw. This method doesn't fit this pattern.
IIUIC, this method attempts to work out whether the given module is a submodule for another project. If this is so, would it be better to call this method something like `InitialiseAsSubmodule`?

. . . 

Also, do you think we could move this method into its own class (that can be injected and tested easier), thus breaking the monolith? E.g.:

```
public sealed class SubmoduleInitialiser : ISubmoduleInitialiser 
{
    public (IGitModule superprojectModule, string submodulePath, string submoduleName) Init(<input params>) { ... }
}
```

This isn't a blocker though (at this stage).

Reply:
Agree it's not quite the usual `Try*` pattern. I was a bit lax there as it's tucked away internally.

Intention is to initialise submodule related properties, only if it is actually a submodule.

Just pushed a commit that renames it to `InitialiseSubmoduleProperties` and documents the purpose. Does that make things clearer?

---

I see two goals for pulling it out:

1. Splitting the monolith: I feel the biggest bang for the buck here will be to split out the operations from the module, and keep things that are intrinsic to the module itself on the class (working directory, submodule properties, index lock tracking, ...). Command formulation and parsing feels pretty separate to me, though of course such things run in the _context_ of a module. That feels like a pretty clear relationship, and something I hope we move towards.

2. Testability: `GitModule` has two external dependencies that have not been mocked/injected so far.  #5281 tackles one (git process execution) and the other is the file system. We have a library that provides `IFileSystem` and I'd like to explore passing in an optional instance to the `GitModule` constructor which would then be used to do this submodule initialisation.

One day I would like to understand why we allow passing invalid paths (`null`, `""`, or `c:\non\existent\path`) for the module's working directory. That smells a bit like a hole in the design to investigate.